### PR TITLE
blobdb migrator for compressing form XML data

### PR DESCRIPTION
**Pull request by snopoke**

### CODE REVIEW

File: corehq/blobs/migrate.py [hunk 1]
```diff
@@ -414,23 +451,26 @@ class BackendMigrator(Migrator):
...
+    def get_document_provider(self, date_range=None, domain=None):
         iteration_key = self.iteration_key
         if date_range:
             (start, end) = date_range
             self.reindexer.date_range = date_range
-            iteration_key = '{}-{}-{}'.format(self.iteration_key, start, end)
+            iteration_key = f'{self.iteration_key}-{start}-{end}'
+        if domain:
+            self.reindexer.domain = domain
+            iteration_key = f'{self.iteration_key}-{domain}'
```
Comment (millerdev) [inline corehq/blobs/migrate.py:472]:
```suggestion
            iteration_key = f'{iteration_key}-{domain}'
```

File: corehq/blobs/migrate.py [hunk 2]
```diff
@@ -414,23 +451,26 @@ class BackendMigrator(Migrator):
...
         return _migrator_with_worker_pool(migrator, self.reindexer, **kw)
 
-    def get_document_provider(self, date_range=None):
+    def get_document_provider(self, date_range=None, domain=None):
         iteration_key = self.iteration_key
         if date_range:
             (start, end) = date_range
             self.reindexer.date_range = date_range
-            iteration_key = '{}-{}-{}'.format(self.iteration_key, start, end)
+            iteration_key = f'{self.iteration_key}-{start}-{end}'
```
Comment (millerdev) [inline corehq/blobs/migrate.py:469]:
```suggestion
            iteration_key = f'{iteration_key}-{start}-{end}'
```

File: corehq/blobs/migrate.py [hunk 3]
```diff
@@ -305,11 +313,35 @@ def migrate(self, doc):
...
+            return True
+        self.total_blobs += 1
+        try:
+            content = meta.open()
+        except NotFound:
+            self.save_backup(doc)
+        else:
+            with content, NamedTemporaryFile() as buffer:
+                shutil.copyfileobj(content, buffer)
+                buffer.seek(0)
```
Comment (millerdev) [inline corehq/blobs/migrate.py:328]: I'm assuming the reason for using `NamedTemporaryFile` is to work around the [`BlobStream` copy blob feature](https://github.com/dimagi/commcare-hq/blob/fc9f450ad7442e715d07a1dede61bfc887c5632c/corehq/blobs/s3db.py#L69). Would it work to use [`io.BufferedReader`](https://docs.python.org/3.7/library/io.html#io.BufferedReader) to stream directly from `content`?
```suggestion
            with content:
                buffer = BufferedReader(content)
```

File: corehq/blobs/migrate.py [hunk 4]
```diff
@@ -500,9 +539,10 @@ class gmigrator:
                 print("done.")
 
 
-MIGRATIONS = {m.slug: m for m in [
-    BackendMigrator("migrate_backend", BlobDbBackendMigrator),
-    BackendMigrator("migrate_backend_check", BlobDbBackendCheckMigrator),
+MIGRATIONS = {m().slug: m for m in [
+    lambda: BackendMigrator("migrate_backend", BlobDbBackendMigrator),
+    lambda: BackendMigrator("migrate_backend_check", BlobDbBackendCheckMigrator),
+    lambda: BackendMigrator("compress_form_xml", BlobDBCompressionMigrator, type_code=CODES.form_xml),
```
Comment (millerdev) [inline corehq/blobs/migrate.py:545]: I don't see why these were converted to lambdas. Can you explain that?

Comment (snopoke) [reply corehq/blobs/migrate.py:545]: If they are statically created then changing any fields on the objects persist for the duration of the runtime. This isn't an issue in a management command but in test cases if there are multiple calls to the same migration then the field values from previous tests persist.

File: corehq/blobs/migrate.py (hunk 3 above)
Comment (snopoke) [reply corehq/blobs/migrate.py:328]: I did consider this (and I'm sure I tried it in a test) but since the blobdb is reading and writing to the same key I wanted to avoid any edge cases where the write would impact the read. I'm not sure about the `S3BlobDB` but the `FilesystemBlobDB` can't read and write to the same file concurrently.

File: corehq/blobs/migrate.py (hunk 1 above)
Comment (millerdev) [reply corehq/blobs/migrate.py:472]: I'm assuming the previous suggestion I left here was resolved by mistake because it looked so much like the other one. I think it still needs to be changed. If not, can you explain why not?
```suggestion
            iteration_key = f'{iteration_key}-{domain}'
```

**Comment by millerdev** [comment]
Ugh, the (I think mistakenly) resolved comment did not reappear when I added another comment. See https://github.com/dimagi/commcare-hq/pull/27297#discussion_r417503557

**Comment by snopoke** [comment]
> Ugh, the (I think mistakenly) resolved comment did not reappear when I added another comment. See #27297 (comment)

Thanks, I thought I'd fixed it but apparently not. Updated now: f2b0351

**Review by millerdev — APPROVED** [review]

**Comment by snopoke** [comment]
Build failed right at the end and unrelated to this PR
